### ISSUE TITLE
Error: Cannot find server build

### ISSUE THREAD

**Issue by arthurgailes** [brillout/vite-plugin-ssr#935]
### Description

`vite build` works in 0.4.80 and breaks in 0.4.89 with the error message below. I'm not able to parse the reasoning:

```

vite-plugin-ssr v0.4.89 pre-rendering HTML...
Error: [@brillout/vite-plugin-import-build@0.2.16][Wrong Usage] Cannot find server build. (Re-)build your app and try again. If you still get this error, then you may need to manually import the server build, see https://github.com/brillout/vite-plugin-import-build#manual-import
    at createErrorWithCleanStackTrace ([mydir]/node_modules/.pnpm/@brillout+vite-plugin-import-build@0.2.16/node_modules/@brillout/vite-plugin-import-build/dist/utils/createErrorWithCleanStackTrace.js:9:15)
    at assertUsage ([mydir]/node_modules/.pnpm/@brillout+vite-plugin-import-build@0.2.16/node_modules/@brillout/vite-plugin-import-build/dist/utils/assert.js:40:92)
    at loadServerBuild ([mydir]/node_modules/.pnpm/@brillout+vite-plugin-import-build@0.2.16/node_modules/@brillout/vite-plugin-import-build/dist/loadServerBuild/index.js:36:29)
    at async loadBuild ([mydir]/node_modules/.pnpm/vite-plugin-ssr@0.4.89_33oe7thcih3clfsuwaxtnji33q/node_modules/vite-plugin-ssr/dist/cjs/node/plugin/plugins/importBuild/loadBuild.js:16:9)
    at async initGlobalContext ([mydir]/node_modules/.pnpm/vite-plugin-ssr@0.4.89_33oe7thcih3clfsuwaxtnji33q/node_modules/vite-plugin-ssr/dist/cjs/node/runtime/globalContext.js:36:30)
    at async runPrerender ([mydir]/node_modules/.pnpm/vite-plugin-ssr@0.4.89_33oe7thcih3clfsuwaxtnji33q/node_modules/vite-plugin-ssr/dist/cjs/node/prerender/runPrerender.js:45:5)
    at async prerender ([mydir]/node_modules/.pnpm/vite-plugin-ssr@0.4.89_33oe7thcih3clfsuwaxtnji33q/node_modules/vite-plugin-ssr/dist/cjs/node/prerender.js:6:5)
    at async triggerFullBuild ([mydir]/node_modules/.pnpm/vite-plugin-ssr@0.4.89_33oe7thcih3clfsuwaxtnji33q/node_modules/vite-plugin-ssr/dist/cjs/node/plugin/plugins/autoFullBuild.js:58:9)
    at async Object.writeBundle ([mydir]/node_modules/.pnpm/vite-plugin-ssr@0.4.89_33oe7thcih3clfsuwaxtnji33q/node_modules/vite-plugin-ssr/dist/cjs/node/plugin/plugins/autoFullBuild.js:24:17)
    at async Promise.all (index 0)
    at async PluginDriver.hookParallel (file://[mydir]/node_modules/.pnpm/rollup@2.79.1/node_modules/rollup/dist/es/shared/rollup.js:22676:9)
    at async file://[mydir]/node_modules/.pnpm/rollup@2.79.1/node_modules/rollup/dist/es/shared/rollup.js:23804:13
    at async catchUnfinishedHookActions (file://[mydir]/node_modules/.pnpm/rollup@2.79.1/node_modules/rollup/dist/es/shared/rollup.js:23126:20)
    at async doBuild (file://[mydir]/node_modules/.pnpm/vite@3.2.4_@types+node@18.0.0/node_modules/vite/dist/node/chunks/dep-67e7f8ab.js:45409:22)
    at async build (file://[mydir]/node_modules/.pnpm/vite@3.2.4_@types+node@18.0.0/node_modules/vite/dist/node/chunks/dep-67e7f8ab.js:45237:16)
    at async CAC.<anonymous> (file://[mydir]/node_modules/.pnpm/vite@3.2.4_@types+node@18.0.0/node_modules/vite/dist/node/cli.js:756:9)
 ELIFECYCLE  Command failed with exit code 1.

```

`vite dev` works in either version. Any assistance you could give would be much appreciated.

**Comment by brillout** [comment]
Are you using Yarn?

**Comment by arthurgailes** [comment]
No, pnpm. In a workspace if that matters.

**Comment by arthurgailes** [comment]
```
vite-plugin-ssr v0.4.89 pre-rendering HTML...
[@brillout/vite-plugin-import-build@0.2.16][DEBUG] DEBUG_LOGS_RUNTIME [begin]
[@brillout/vite-plugin-import-build@0.2.16][DEBUG] process.platform "linux"
[@brillout/vite-plugin-import-build@0.2.16][DEBUG] process.release {"name":"node","lts":"Gallium","sourceUrl":"https://nodejs.org/download/release/v16.16.0/node-v16.16.0.tar.gz","headersUrl":"https://nodejs.org/download/release/v16.16.0/node-v16.16.0-headers.tar.gz"}
[@brillout/vite-plugin-import-build@0.2.16][DEBUG] navigator undefined
[@brillout/vite-plugin-import-build@0.2.16][DEBUG] cwd /[workspaceRoot]/apps/hilton/vite
[@brillout/vite-plugin-import-build@0.2.16][DEBUG] importer.status SET
[@brillout/vite-plugin-import-build@0.2.16][DEBUG] importer.paths.autoImporterFilePathOriginal /[workspaceRoot]/node_modules/.pnpm/@brillout+vite-plugin-import-build@0.2.16/node_modules/@brillout/vite-plugin-import-build/dist/autoImporter.js
[@brillout/vite-plugin-import-build@0.2.16][DEBUG] importer.paths.autoImporterFileDirActual /[workspaceRoot]/node_modules/.pnpm/@brillout+vite-plugin-import-build@0.2.16/node_modules/@brillout/vite-plugin-import-build/dist
[@brillout/vite-plugin-import-build@0.2.16][DEBUG] importer.paths.importBuildFilePathRelative ../../../../../../../dist/apps/hilton/vite-ssr/server/importBuild.cjs
[@brillout/vite-plugin-import-build@0.2.16][DEBUG] importer.paths.importBuildFilePathOriginal /[workspaceRoot]/dist/apps/hilton/vite-ssr/server/importBuild.cjs
[@brillout/vite-plugin-import-build@0.2.16][DEBUG] importer.paths.importBuildFilePathResolved() /[workspaceRoot]/dist/apps/hilton/vite-ssr/server/importBuild.cjs
[@brillout/vite-plugin-import-build@0.2.16][DEBUG] requireError undefined
[@brillout/vite-plugin-import-build@0.2.16][DEBUG] success false
[@brillout/vite-plugin-import-build@0.2.16][DEBUG] DEBUG_LOGS_RUNTIME [end]
Error: [@brillout/vite-plugin-import-build@0.2.16][Wrong Usage] Cannot find server build. (Re-)build your app and try again. If you still get this error, then you may need to manually import the server build, see https://github.com/brillout/vite-plugin-import-build#manual-import
    at createErrorWithCleanStackTrace (/[workspaceRoot]/node_modules/.pnpm/@brillout+vite-plugin-import-build@0.2.16/node_modules/@brillout/vite-plugin-import-build/dist/utils/createErrorWithCleanStackTrace.js:9:15)
    at assertUsage (/[workspaceRoot]/node_modules/.pnpm/@brillout+vite-plugin-import-build@0.2.16/node_modules/@brillout/vite-plugin-import-build/dist/utils/assert.js:40:92)
    at loadServerBuild (/[workspaceRoot]/node_modules/.pnpm/@brillout+vite-plugin-import-build@0.2.16/node_modules/@brillout/vite-plugin-import-build/dist/loadServerBuild/index.js:36:29)
    at async loadBuild (/[workspaceRoot]/node_modules/.pnpm/vite-plugin-ssr@0.4.89_33oe7thcih3clfsuwaxtnji33q/node_modules/vite-plugin-ssr/dist/cjs/node/plugin/plugins/importBuild/loadBuild.js:16:9)
    at async initGlobalContext (/[workspaceRoot]/node_modules/.pnpm/vite-plugin-ssr@0.4.89_33oe7thcih3clfsuwaxtnji33q/node_modules/vite-plugin-ssr/dist/cjs/node/runtime/globalContext.js:36:30)
    at async runPrerender (/[workspaceRoot]/node_modules/.pnpm/vite-plugin-ssr@0.4.89_33oe7thcih3clfsuwaxtnji33q/node_modules/vite-plugin-ssr/dist/cjs/node/prerender/runPrerender.js:45:5)
    at async prerender (/[workspaceRoot]/node_modules/.pnpm/vite-plugin-ssr@0.4.89_33oe7thcih3clfsuwaxtnji33q/node_modules/vite-plugin-ssr/dist/cjs/node/prerender.js:6:5)
    at async triggerFullBuild (/[workspaceRoot]/node_modules/.pnpm/vite-plugin-ssr@0.4.89_33oe7thcih3clfsuwaxtnji33q/node_modules/vite-plugin-ssr/dist/cjs/node/plugin/plugins/autoFullBuild.js:58:9)
    at async Object.writeBundle (/[workspaceRoot]/node_modules/.pnpm/vite-plugin-ssr@0.4.89_33oe7thcih3clfsuwaxtnji33q/node_modules/vite-plugin-ssr/dist/cjs/node/plugin/plugins/autoFullBuild.js:24:17)
    at async Promise.all (index 0)
    at async PluginDriver.hookParallel (file:///[workspaceRoot]/node_modules/.pnpm/rollup@2.79.1/node_modules/rollup/dist/es/shared/rollup.js:22676:9)
    at async file:///[workspaceRoot]/node_modules/.pnpm/rollup@2.79.1/node_modules/rollup/dist/es/shared/rollup.js:23804:13
    at async catchUnfinishedHookActions (file:///[workspaceRoot]/node_modules/.pnpm/rollup@2.79.1/node_modules/rollup/dist/es/shared/rollup.js:23126:20)
    at async doBuild (file:///[workspaceRoot]/node_modules/.pnpm/vite@3.2.4_@types+node@18.0.0/node_modules/vite/dist/node/chunks/dep-67e7f8ab.js:45409:22)
    at async build (file:///[workspaceRoot]/node_modules/.pnpm/vite@3.2.4_@types+node@18.0.0/node_modules/vite/dist/node/chunks/dep-67e7f8ab.js:45237:16)
    at async CAC.<anonymous> (file:///[workspaceRoot]/node_modules/.pnpm/vite@3.2.4_@types+node@18.0.0/node_modules/vite/dist/node/cli.js:756:9)
 ELIFECYCLE  Command failed with exit code 1.

```

**Comment by brillout** [comment]
The problem seems to be that the `dist/` directory isn't a child/descendent directory of `process.cwd()`.

What's your `config.build.outDir`? Are you moving the `dist/` directory?

**Comment by arthurgailes** [comment]
Yes. In vite.config.ts:
```
const root = __dirname; // project root
const workspaceRoot = `${__dirname}/../../../`;

...
{build: {
      outDir: `${workspaceRoot}/dist/apps/hilton/vite-ssr`,
    }}

**Comment by brillout** [comment]
I see. Follow https://github.com/brillout/vite-plugin-import-build#manual-import and let me know if that works.

**Comment by arthurgailes** [comment]
Got it. Yes, this works, as does removing build.outDir. Thanks very much for your help.

**Comment by brillout** [comment]
👍 I guess we can close then.

> I'm not able to parse the reasoning:

What is it you found difficult to comprehend?

**Comment by arthurgailes** [comment]
I'm sorry, this actually doesn't quite work. If I remove build.outDir, it works, and it will continue to build if I restore and make these changes. But internally, it's somehow still referencing the default /dist directory and reading from there. If I delete the default /dist directory, it breaks again.

Here's a simplified version of my code:
```
import express from 'express';
import { join } from 'path';
import compression from 'compression';
import { createServer as createViteServer } from 'vite';
import { renderPage } from 'vite-plugin-ssr/server';
import sirv from 'sirv';
import { config } from 'dotenv';

const isProduction = process.env.NODE_ENV === 'production';

const root = `${__dirname}/..`;
const workspaceRoot = join(root, `/../../..`);
const outDir = join(workspaceRoot, "dist/apps/my_app")

startServer();

async function startServer() {
  const app = express();
  
  app.use(compression());
  if (isProduction) {
    await import(`${workspaceRoot}/${outDir}/server/importBuild.cjs`);
    app.use(sirv(`${workspaceRoot}/${outDir}/client`));
  } else {
    const viteDevMiddleware = (
      await createViteServer({
        root: `${root}`, ///../../../dist/apps/hilton/vite-ssr
        server: { middlewareMode: true },
        configFile: `${root}/vite.config.ts`,
      })
    ).middlewares;
    app.use(viteDevMiddleware);
  }

  app.get('*', async (req, res, next) => {
    const urlOriginal = req.url;
    const pageContextInit = {
      urlOriginal,
    };
    const pageContext = await renderPage(pageContextInit);
    const { httpResponse } = pageContext;

    // accepts OPTIONS
    if (req.method === 'OPTIONS') {
      res.writeHead(200);
      return res.end();
    }

    if (!httpResponse) return next();
    else {
      const { pipe, earlyHints, contentType, statusCode } = httpResponse;
      if (res.writeEarlyHints)
        res.writeEarlyHints({ link: earlyHints.map((e) => e.earlyHintLink) });
      // res.status(statusCode).type(contentType);

      pipe(res);
    }
  });

  const port = process.env.port || 3000;
  app.listen(port);
  console.log(`Server running at http://localhost:${port}`);
}

```

**Comment by brillout** [comment]
> What is it you found difficult to comprehend?

Also, this would help improve the docs.

**Comment by arthurgailes** [comment]
Fun. Here you go: https://github.com/arthurgailes/vite-plugin-ssr-workspace-repro

ssr({prerender:true}) seems to be the causal factor.

**Comment by brillout** [comment]
👍 I could reproduce. I think I know what's going on, I'll try to fix it later today.

**Comment by brillout** [comment]
Fix pre-released in `0.4.131-commit-ec8b2f39b`.

**Comment by arthurgailes** [comment]
Still broken on windows. Repro updated. 

```
Debug info (this is for the vite-plugin-ssr maintainers; you can ignore this): Not a posix path: C:\[...mypath]\vite-plugin-ssr-workspace-repro\packages\boilerplate-react-ts/../../dist/boilerplate-react-ts
    at assertPosixPath (C:\[...mypath]\vite-plugin-ssr-workspace-repro\node_modules\.pnpm\vite-plugin-ssr@0.4.132_vite@4.3.9\node_modules\vite-plugin-ssr\dist\cjs\utils\filesystemPathHandling.js:23:25)
    at determineOutDir (C:\[...mypath]\vite-plugin-ssr-workspace-repro\node_modules\.pnpm\vite-plugin-ssr@0.4.132_vite@4.3.9\node_modules\vite-plugin-ssr\dist\cjs\utils\getOutDirs.js:28:50)
    at config (C:\[mypath]\vite-plugin-ssr-workspace-repro\node_modules\.pnpm\vite-plugin-ssr@0.4.132_vite@4.3.9\node_modules\vite-plugin-ssr\dist\cjs\node\plugin\plugins\buildConfig.js:35:57)
    at runConfigHook (file:///C:/[...mypath]/vite-plugin-ssr-workspace-repro/node_modules/.pnpm/vite@4.3.9_@types+node@18.16.16/node_modules/vite/dist/node/chunks/dep-e8f070e8.js:64543:31)
    at resolveConfig (file:///C:/[...mypath]/vite-plugin-ssr-workspace-repro/node_modules/.pnpm/vite@4.3.9_@types+node@18.16.16/node_modules/vite/dist/node/chunks/dep-e8f070e8.js:64026:14)
    at build (file:///C:/[...mypath]/vite-plugin-ssr-workspace-repro/node_modules/.pnpm/vite@4.3.9_@types+node@18.16.16/node_modules/vite/dist/node/chunks/dep-e8f070e8.js:46318:20)
    at CAC.<anonymous> (file:///C:/[...mypath]/vite-plugin-ssr-workspace-repro/node_modules/.pnpm/vite@4.3.9_@types+node@18.16.16/node_modules/vite/dist/node/cli.js:812:9)
```

**Comment by brillout** [comment]
Let me know if `0.4.132-commit-3f235bc84` works.

**Comment by arthurgailes** [comment]
```
Debug info (this is for the vite-plugin-ssr maintainers; you can ignore this): Not a posix path: C:\[...mydir]\vite-plugin-ssr-workspace-repro\packages\boilerplate-react-ts/../../dist/boilerplate-react-ts
    at assertPosixPath (C:\[...mydir]\vite-plugin-ssr-workspace-repro\node_modules\.pnpm\vite-plugin-ssr@0.4.132-commit-3f235bc84_vite@4.3.9\node_modules\vite-plugin-ssr\dist\cjs\utils\filesystemPathHandling.js:23:25)
    at getOutDirs_prerender (C:\[...mydir]\vite-plugin-ssr-workspace-repro\node_modules\.pnpm\vite-plugin-ssr@0.4.132-commit-3f235bc84_vite@4.3.9\node_modules\vite-plugin-ssr\dist\cjs\utils\getOutDirs.js:20:50)
    at runPrerender (C:\[...mydir]\vite-plugin-ssr-workspace-repro\node_modules\.pnpm\vite-plugin-ssr@0.4.132-commit-3f235bc84_vite@4.3.9\node_modules\vite-plugin-ssr\dist\cjs\node\prerender\runPrerender.js:37:75)
    at triggerFullBuild (C:\[...mydir]\vite-plugin-ssr-workspace-repro\node_modules\.pnpm\vite-plugin-ssr@0.4.132-commit-3f235bc84_vite@4.3.9\node_modules\vite-plugin-ssr\dist\cjs\node\plugin\plugins\autoFullBuild.js:80:9)
t-3f235bc84_vite@4.3.9\node_modules\vite-plugin-ssr\dist\cjs\node\plugin\plugins\autoFullBuild.js:30:25)
    at async Promise.all (index 0)
    at PluginDriver.hookParallel (file:///C:/[...mydir]/vite-plugin-ssr-workspace-repro/node_modules/.pnpm/rollup@3.23.1/node_modules/rollup/dist/es/shared/node-entry.js:24590:9)
    at file:///C:/[...mydir]/vite-plugin-ssr-workspace-repro/node_modules/.pnpm/rollup@3.23.1/node_modules/rollup/dist/es/shared/node-entry.js:25954:13
    at catchUnfinishedHookActions (file:///C:/[...mydir]/vite-plugin-ssr-workspace-repro/node_modules/.pnpm/rollup@3.23.1/node_modules/rollup/dist/es/shared/node-entry.js:25028:16)
    at build (file:///C:/[...mydir]/vite-plugin-ssr-workspace-repro/node_modules/.pnpm/vite@4.3.9_@types+node@18.16.16/node_modules/vite/dist/node/chunks/dep-e8f070e8.js:46477:22)
    at CAC.<anonymous> (file:///C:/[...mydir]/vite-plugin-ssr-workspace-repro/node_modules/.pnpm/vite@4.3.9_@types+node@18.16.16/node_modules/vite/dist/node/cli.js:812:9)
```

**Comment by brillout** [comment]
Interesting, it does seem to be a Vite bug after all. I'll have a look at it later.

**Comment by brillout** [comment]
Fix released in `0.4.132-commit-9ae9aa2`. I've tested it; it should work for you now. But let me know if you still run into any issue.

> I'm not able to parse the reasoning

I've improved the readme of https://github.com/brillout/vite-plugin-import-build. Should be less convoluted now.

**Comment by arthurgailes** [comment]
Verified this works, thank you.
